nemo/collections/asr/data/audio_to_text.py

```diff
@@ -208,6 +226,125 @@ def expand_audio_filepaths(audio_tar_filepaths, shard_strategy: str, world_size:
+                                result = list(tqdm(p.imap(store_object_get, audio_objects), total=len(audio_objects)))
+                        else:
+                            logging.debug('Using a single process.')
+                            result = []
+                            for audio_object in tqdm(audio_objects):
+                                result.append(audio_object.get() is not None)
+
+                        if any(result) == False:
+                            raise RuntimeError('Some files not downloaded successfully')
+                        logging.info('Caching complete')
+
+                else:
+                    # Nothing to do here
+                    logging.debug('Manifest is not on a store: %s', manifest_file)
+
+        # Handle distributed environment
+        if shared_cache is None:
+            shared_cache = is_store_cache_shared()
+
+        if shared_cache:
+            logging.debug('Cache is shared among nodes.')
+            is_rank_zero = is_global_rank_zero()
+        else:
+            logging.debug('Cache is not shared among nodes.')
+            local_rank = int(os.environ.get("LOCAL_RANK", 0))
+            is_rank_zero = local_rank == 0
+
+        if torch.distributed.is_available() and torch.distributed.is_initialized():
+            logging.debug('Distributed environment is available and initialized.')
+            if is_rank_zero:
+                logging.info('Cache data from %s rank 0', 'global' if shared_cache else 'local')
+                cache_data(
+                    manifest_filepaths=manifest_filepaths,
+                    cache_audio=cache_audio,
+                    num_workers=num_workers,
+                    max_num_workers=max_num_workers,
+                )
+            logging.debug('Reached barrier')
+            torch.distributed.barrier()
+
+        elif is_rank_zero:
+            logging.warning(
+                'Torch distributed is not initialized and caching may be prone to data race conditions. '
+                'Now caching data from %s rank 0. If other ranks pass this '
+                'before rank 0, errors might result.',
+                'global' if shared_cache else 'local',
+            )
+            cache_data(
+                manifest_filepaths=manifest_filepaths,
+                cache_audio=cache_audio,
+                num_workers=num_workers,
+                max_num_workers=max_num_workers,
+            )
+
+        else:
+            logging.warning(
+                'Torch distributed is not initialized and caching may be prone to data race conditions. '
+                'This process is not %s rank 0, and now will go to sleep for %d seconds. If this '
+                'rank wakes from sleep prior to rank 0 finishing downloading, errors might result.',
+                'global' if shared_cache else 'local',
+                num_store_manifests * sleep_seconds,
+            )
+            time.sleep(num_store_manifests * sleep_seconds)
+
```

Comment:
Is this the busy waiting part? Doesnt num store manifest * sleep seconds become dozens seconds?